### ISSUE TITLE
Missing 'Albums' list in the Info/NowPlaying view (ref. #1460)

### ISSUE THREAD

**Issue by nousernouser** [CDrummond/cantata#1461]
Missing 'Albums' list in the Info/NowPlaying view (ref. #1460)

Hi 

As shown in these screenshots:

<https://github.com/CDrummond/cantata/issues/1460#issuecomment-468836120>

there is no 'Albums' list when using 'Composer' for any genre set in the "Composer Support" tweak.

I would be very grateful if you wanted to add this very useful feature too.

Sorry for the trouble and thank you very much again!!!

**Comment by CDrummond** [comment]
Try latest git/master

**Comment by nousernouser** [comment]
Sorry for the delay.
Tried, but still miss the 'Album' list...

**Comment by CDrummond** [comment]
I've just tried this:

1. Made a copy of 3 albums from the same artist
2. Set the 'Composer' tag of 2 of the albums, and set their genre to Classical
3. Refreshed Cantata's DB
4. Played a track from one of the edited albums
5. I Info view I see the other 2 albums listed, and I can navigate into

So, sorry, but it is working.

**Comment by nousernouser** [comment]
Il giorno Sat, 09 Mar 2019 01:18:09 -0800
CraigD <notifications@github.com> ha scritto:

> I've just tried this:
> 
> 1. Made a copy of 3 albums from the same artist
> 2. Set the 'Composer' tag of 2 of the albums, and set their genre to Classical
> 3. Refreshed Cantata's DB
> 4. Played a track from one of the edited albums
> 5. I Info view I see the other 2 albums listed, and I can navigate into
> 
> So, sorry, but it is working.


Hi

First of all, thank you for the effort. 

You did exactly the same steps I did!... but I cannot see the other 2 albums
listed :(

Did you also set the 'Composer support' tweak to 'Classical' as I did?

And, please, could you tell me if you had metadata embedded into tracks or
coming from playlist files for the 3 albums?

And, please, could you tell me which were 'Artist', 'Performer' and 'Composer'
for one of the tracks (say the 1st) of each of the 3 albums?

Thanks again
Regards
A.

**Comment by CDrummond** [comment]
Yes, I set the genre to "Classical" and set this in the tweaks page. All meta-data is in the files, I used kid3-qt to edit these. No "Performer", Artist (& Album Artist) was "Iron Maiden", and I made up a string for Composer.

**Comment by nousernouser** [comment]
Well... I did other tests, and I have found is that the problem seems to be related to 'Performer'.

To reproduce: 

- 'Composer' set in the "Composer Support" tweak.
- 3 albums, genre Classical, same artist/composer
    - 1 (say album **A**) with tracks where the performer tag is set DIFFERENTLY from artist/composer (please note that this is a quite common situation with classical music)
    - 1 (say album **B**) with tracks where the performer tag is set the SAME OF artist/composer
    - 1 (say album **C**) with tracks where the performer tag is NOT set

(see the attached screenshots)

In the now playing (info) view, when playing a track either from album A or B or C, **only** the album A is listed (being album B and album C not listed).

So it seems to be an issue that presents itself only in these circumstances, which however are very common for those who also listen to classical albums!

I hope this can be helpful in identifying the problem, and I hope you will kindly work on it.

Thank you a lot!
Regards

![albumABC](https://user-images.githubusercontent.com/35839588/54078062-0b2b8180-42c2-11e9-8f05-3b30daa42494.png)

![infoview](https://user-images.githubusercontent.com/35839588/54078067-18e10700-42c2-11e9-8d13-368830551387.png)

**Comment by nousernouser** [comment]
(just updated wrong screenshot)

**Comment by CDrummond** [comment]
Tried, still working for me.

When searching for these albums, Cantata uses its cache of the MPD database. This is stored as an SQLite file. Cantata does not store the **Performer** field, only Artist, AlbumArtist, and Composer. Therefore, **Performer** cannot have any affect on this.

To find these albums, Cantata performs the following queries:

First, searches for artist match. Here it _should_ be checking for artist = composer.
```
sqlite3 -column -header ~/.local/share/cantata/library/_home_craig_.local_share_cantata_mpd_socket.sql "select distinct album, albumId, albumSort from songs where artist = 'Blobby'"
```

Then, if no albums are found it tries composer:

```
sqlite3 -column -header ~/.local/share/cantata/library/_home_craig_.local_share_cantata_mpd_socket.sql "select distinct album, albumId, albumSort from songs where composer = 'Blobby'"
```

If you want to try these on your DB, replace `~/.local/share/cantata/library/_home_craig_.local_share_cantata_mpd_socket.sql` with the correct file on your system, and replace `Blobby` with the correct composer.

**Comment by nousernouser** [comment]
Il giorno Sun, 10 Mar 2019 00:25:20 -0800
CraigD <notifications@github.com> ha scritto:

> Tried, still working for me.

not for me :(

> When searching for these albums, Cantata uses its cache of the MPD database.
> This is stored as an SQLite file. Cantata does not store the **Performer**
> field, only Artist, AlbumArtist, and Composer. Therefore, **Performer**
> cannot have any affect on this.

I am not sure about this, as I have Performer and it is "assigned" to Artist
and, if I have well understood, you did the tests without set the Performer
tag...

> To find these albums, Cantata performs the following queries:
> [...]

Thanks for these useful information...

Let me do some more investigation...

Regards

**Comment by nousernouser** [comment]
Here I am, sorry for the delay but I had work commitments.

However, I did a lot of testing trying to figure out the reason for the difference between the two cases, mine and yours.

I used both (1) albums with separate music tracks and embedded tags, and (2) albums with single files and CUE playlists.

The conclusion is that everything "works" as expected in the case of albums (1), while albums (2) "disappear" (and I think this explains the difference in behavior between our two cases, as I have both kind of albums in my music db).

Going deeper into the question (I examined the `mpd-interface/cuefile.cpp` file), it turns out that the problem is related to tag management during the cue file parsing.

Very briefly, `artist` should match` performer` but, if `genre` is one in the "Composer Support" tweak list, then it should match` composer` instead.

In this sense I modified the code of the `mpd-interface/cuefile.cpp` file, and rebuild Cantata: all works fine!

I would therefore like to share the code with the changes I have made, but I still need to perfect the way to determine the tag to associate with `artist`... For now I simply used the check:

~~~
if (genre == "Classical") { 
...
} else {
...
}
~~~

but this obviously is not enough: I need to check if `genre` is one in the" Composer Support "tweak list.

So, please help me a little (I am not so expert with cpp and the Cantata code): you could kindly tell me the right code to be used in `mpd-interface/cuefile.cpp` to check inside `CueFile::parse` if `genre` is on the list set by the user in the "Composer Support" tweak?

Thanks a lot
Regards

**Comment by CDrummond** [comment]
I'll re-open this issue.

As for the code change, something like:

```
if (Song::composerGenres().contains(genre)) {
} else {
}
```

should work.

**Comment by nousernouser** [comment]
Hi. First of all: thanks for having re-opened, and for the very useful suggestion on the code.

I modified the code of the `mpd-interface/cuefile.cpp` file, and rebuild Cantata. Then tested ...

Now, `artist` is set and it matches `performer` but, if `genre` is one in the *"Composer Support"* tweak list, then it matches `composer` instead ... And this is just the expected behavior ...

So now it works!!!

I attach the `cuefile.cpp` file containing the changes, as well as the `cuefile.cpp.patch` file (referring to the sources updated to commit 985f5e6).

As you can see, I allowed myself to introduce some further small improvements to improve support for CUE files. I also tried the handling for the "`REM ORIGINALDATE`" CUE tag (mpd: `originalDate`): no building errors, but it doesn't seem to work (I think due to my insufficient understanding of the Cantata code and/or C++) ... Moreover, there would also be something to manage the tag "`REM LABEL`" (mpd: `label`) ... If you can help me, I will try to fix them ...

Thanks again very much

Regards


[cuefile.zip](https://github.com/CDrummond/cantata/files/3005460/cuefile.zip)

**Comment by CDrummond** [comment]
I've applied (as slightly modified version of) you changes to git/master.

As for 'originalDate' - the code looks OK, so not sure why it is not working. I've added originalDate to the debug statement. Does it appear there?

As for label - there is no point to parsing this, it will not be used by Cantata.

**Comment by nousernouser** [comment]
Hi

> I've applied (as slightly modified version of) you changes to git/master.

Great! Thank you (also for the good fixes to the code changes).


> As for 'originalDate' - the code looks OK, so not sure why it is not working. I've added originalDate to the debug statement. Does it appear there?

Yes. It appears in the debug,  but is missing from the Metadata list shown in the 'now playing view'.

More, there are other tags which are missing... I have set the same tags either through a cue file (single file album) and embedded (separated file tracks album), both with genre being in the composer tweak list: some of the cue tags which seems to me being supported by the cantata code, are not shown in the 'now playing view'. 

To contribute to debugging, I describe below some things that result from the tests - please have a look to the screenshots at the bottom (cue tags: ALBUM-C1 red cover; embedded tags: ALBUM-E1 black cover)

- grouping albums is working as expected, either with genre in the composer tweak list or not

- artist, composer and performer are correctly shown with ALBUM-E1, not with ALBUM-C1 (despite they are correctly assigned by cuefile.cpp as results by the debug statement)

- originaldate, discnumber, disctotal, tracktotal (which seems to me being supported by the cantata code) are missing with ALBUM-C1 (originaldate appears correcly by the debug statement)

- other "missing" tags with ALBUM-C1: I suppose they are missing simply because not handled by cuefile.cpp

I am available for testing and whatever else is needed. If you want I am also available to further modify the code to introduce the missing cue tags (for this I believe that I need some more information from you on how to proceed...).

--

Furthermore, I realized that I had made an evaluation error regarding the originalDate tag. In fact, originalDate is different from originalYear: MusicBrainz defines both, but originalYear is properly a year (YYYY), while originalDate is a date (YYYY-MM-DD); moreover it is not clear how MPD considers the originalDate tag.

Since Cantata treats this tag as a year (origYear variable) and this is still satisfactory, it is simply a matter of replacing originalDate with originalYear in the code and in the comments of cuefile.cpp - to fix it (and other little typos), I have already made these changes: I attach the correct file (build and tested ok). Sorry for my mistake.

Thanks
Regards

[cuefile.cpp.zip](https://github.com/CDrummond/cantata/files/3024822/cuefile.cpp.zip)

**ALBUM-C1 (cue tags)**
![ALBUM-C1](https://user-images.githubusercontent.com/35839588/55266828-31b85900-527f-11e9-9b39-0a085a2717d1.png)

**ALBUM-E1 (embedded tags)**
![ALBUM-E1](https://user-images.githubusercontent.com/35839588/55266839-3977fd80-527f-11e9-8bb5-d89c9e92e748.png)

**Comment by CDrummond** [comment]
The metadata view does not parse Cue files - hence why it is lacking.

As to the changes to the cue file reading, I'm now not sure these are correct. You're setting album-artist to the composer/albumComposer. This should not be handled here. The composer should be set to the composer. Cantata then has code (Song::albumArtistOrComposer) which checks which one to use based upon the tweaks setting. Also, Cantata has no concept of 'Album Composer' for any other tracks.

I'm thinking of reverting:

1. Reading of 'Album Composer' - is not used anywhere else.
2. Setting AlbumArtist to the composer value. This should not be done here
3. The various artists checking, not sure why VA is being 'ignored'

**Comment by nousernouser** [comment]
> The metadata view does not parse Cue files - hence why it is lacking.

Ah ok... Just for my curiosity: which are the code file involved here?

> As to the changes to the cue file reading, I'm now not sure these are correct. [...]
> I'm thinking of reverting:
>   Reading of 'Album Composer' - is not used anywhere else.
>   Setting AlbumArtist to the composer value. This should not be done here
>   The various artists checking, not sure why VA is being 'ignored'

Please, can you wait some days? 

I guess test this: revert changes, buil and test what happens...

With regard to VA -- it not is (should not be) 'ignored'... For a VA album:
- when a song has artist (performer or composer) the song should keep it (artist sould not being reassigned to be VA)
- all the songs in the album should have albumartist assignet to VA
I guess also check this...

Updates coming soon...

Regards

**Comment by CDrummond** [comment]
The metadata view does not parse Cue files - hence why it is lacking.

    Ah ok... Just for my curiosity: which are the code file involved here?

`context/songview.cpp` `SongView::loadMetadata()` gets meta-data from tags. I guess you could add a 'if (cue file)' here...

**Comment by nousernouser** [comment]
not gone ... only too much to do ... but working ... please give me some more time...

**Comment by CDrummond** [comment]
Take whatever time you need, there's no rush - I have no plans for a release at the moment.

**Comment by nousernouser** [comment]
Hi, a bit of news! I try to go with order and be brief as much as possible...

A premise. To discuss appropriately the reverts you had announced, and to try to resolve the issue of the "missing" tags (see previous messages), I made several attempts to modify the code (`cuefile.cpp`, `songview.cpp`), with related debugging and lot of tests -- which, moreover, led me to deepen my understanding of the Cantata code (`librarydb.cpp`, `mpdparseutils.cpp` etc.) and learn a bit more of Qt.

Doing such attempts, I ended up reviewing and modifying the `cuefile.cpp` and `songview.cpp` code (please, see datailed comments and clarifications inside the attached files):

### mpd-interface/cuefile.cpp

- According to you, *Album Artist* (which is not used) no longer exists anywhere.

- Again in agreement with you, setting *Album Artist* to the *Performer* or *Composer* should not be done here, and indeed it is no longer done.

- Instead, the VA checking is necessary to be able to manage correctly some track tags even with VA albums (please, see detailed comments inside the attached code).

- CUE *Albums* (VA albums included) are correctly listed and grouped, both in the Library and the Info (NowPlaying) view, either if using or not the "Composer Support" tweak: **this should finally close the original Cantata issue [#1461](https://github.com/CDrummond/cantata/issues/1461)**!!!.

- The different management of some variables (as can be seen by examining the proposed code), should have led to a greater speed of execution when doing "Refresh Database" (especially appreciable with a large number of audio CUE files, as in my case).

- The way the parsing of CUE tags is managed also allows you to easily add new ones, or use multiple CUE tags at the same MPD tag. Furthermore, all the tricky forms in which tags are used with `REM` (eg `REM COMPOSER ComposerName`), as well as forms with non-standard CUE commands (eg `ORIGINALDATE 1978`), are recognized "automagically".
In fact, the `comment` tag also, which is related to` REM` or `REM COMMENT`, is now managed; multi-line comments are allowed (accordingly with the CUE specification); and the `originaldate` tag is now related to both `ORIGINALDATE` and `ORIGINALYEAR`.

- Moreover, the different parsing flow proposed, should work (as it turns out based on the tests I did with my CUE files) even in the case of CUE files with multiple audio files each with one or multiple tracks: **this should also solve the Cantata issue [#1479](https://github.com/CDrummond/cantata/issues/1479)**!!!


### context/songview.cpp

- Changes in `SongView::loadMetadata()` (section: `if (tags.isEmpty()) {...}`) to match `cuefile.cpp` changes (please, see comments inside the attached code)  
-- *Note: I commented out the previous code of this section because I couldn't find the right way to verify (from inside `songview.cpp`) if the current song is from a CUE file...*


### things remain pending...

Something "strange" happens with `origYear`, `disc` and `comment`  in `SongView::loadMetadata()`. I have done a lot of debugging on this, and now I try to explain the situation, by tracing the metadata of a track from a CUE file.

Consider ***this track in this CUE file*** I have used for testing (I report here only the relevant lines): 
~~~.text
REM GENRE "Classical"
REM DATE "2007-05-08"
REM ORIGINALYEAR "1999"
REM DISCNUMBER "2"
REM "Remark1 - A standard REM."
REM COMMENT "Remark2 - A tricky REM."
REM COMPOSER "H Composer"
PERFORMER "H Performer"
TITLE "ALBUM-C1-HCP"
FILE "audiocd.flac" WAVE
  TRACK 01 AUDIO
    PERFORMER "Performer T1"
    TITLE "TRACK 01"
    PREGAP 00:00:33
    INDEX 01 00:00:00
  ...
~~~

CUE ***file parsing*** is correctly done by `cuefile.cpp`, and metadata values are correctly assigned to the song, as shown by the debug log:

~~~.text
...
2019-05-03 11:01:04 - MPDParseUtils Parsing cue file: "TEST-ALBUM-C1-HCP/audiocd.cue" mpdDir: "/media/AUDIO/AudioHiFi/"
...
2019-05-03 11:01:04 - CueFile FILE   | file: "audiocd.flac" | fileType: "WAVE" | fileValid? true
2019-05-03 11:01:04 - CueFile HEADER | genre: "Classical" | isGenreInComposerSupport? true
2019-05-03 11:01:04 - CueFile        | album: "ALBUM-C1-HCP" | discNo: "2" | year: "2007" | originalYear: "1999"
2019-05-03 11:01:04 - CueFile        | headerPerformer: "H Performer" | headerComposer: "H Composer"
2019-05-03 11:01:04 - CueFile        | REM: "Remark1 - A standard REM. Remark2 - A tricky REM."
...
2019-05-03 11:01:04 - CueFile TRACK -- tCount: 0
2019-05-03 11:01:04 - CueFile        | file: "audiocd.flac"
2019-05-03 11:01:04 - CueFile        | trackNo: "01" | type: "audio" | index: "00:00:00"
2019-05-03 11:01:04 - CueFile        | title: "TRACK 01" | artist: "H Performer" | performer: "H Performer" | composer: "H Composer"
...
2019-05-03 11:01:04 - CueFile SONG  | file: "cue:///TEST-ALBUM-C1-HCP/audiocd.cue?pos=0" | songFile: "TEST-ALBUM-C1-HCP/audiocd.flac"
2019-05-03 11:01:04 - CueFile       | genre: "Classical" | album: "ALBUM-C1-HCP" | year: 2007 | originalYear: 1999
2019-05-03 11:01:04 - CueFile       | discNo: 2 | trackNo: 1 | title: "TRACK 01"
2019-05-03 11:01:04 - CueFile       | artist: "H Performer" | performer: "H Performer" | composer: "H Composer"
2019-05-03 11:01:04 - CueFile       | comment: "Remark1 - A standard REM. Remark2 - A tricky REM."
...
~~~

Debugging of `mpdparseutils` shows metadata are still correct when executing `MPDParseUtils::parseDirItems` (2 passes):

~~~.text
2019-05-03 11:01:04 - MPDParseUtils Parsed cue file, songs: 9 files: QSet("TEST-ALBUM-C1-HCP/audiocd.flac")
2019-05-03 11:01:04 - MPDParseUtils *0*void MPDParseUtils::parseDirItems -- s.
2019-05-03 11:01:04 - MPDParseUtils    file: "cue:///TEST-ALBUM-C1-HCP/audiocd.cue?pos=0"
2019-05-03 11:01:04 - MPDParseUtils    genre: "Classical" | album: "ALBUM-C1-HCP" | year: 2007 | originalYear: 1999
2019-05-03 11:01:04 - MPDParseUtils    discNo: 2 | trackNo: 1 | title: "TRACK 01" | duration:  238
2019-05-03 11:01:04 - MPDParseUtils    artist: "H Performer" | performer: "H Performer" | composer: "H Composer"
2019-05-03 11:01:04 - MPDParseUtils    comment: "Remark1 - A standard REM. Remark2 - A tricky REM."
...
2019-05-03 11:01:04 - MPDParseUtils *1*void MPDParseUtils::parseDirItems -- s.
2019-05-03 11:01:04 - MPDParseUtils    file: "cue:///TEST-ALBUM-C1-HCP/audiocd.cue?pos=0"
2019-05-03 11:01:04 - MPDParseUtils    genre: "Classical" | album: "ALBUM-C1-HCP" | year: 2007 | originalYear: 1999
2019-05-03 11:01:04 - MPDParseUtils    discNo: 2 | trackNo: 1 | title: "TRACK 01" | duration:  238
2019-05-03 11:01:04 - MPDParseUtils    artist: "H Performer" | performer: "H Performer" | composer: "H Composer"
2019-05-03 11:01:04 - MPDParseUtils    comment: "Remark1 - A standard REM. Remark2 - A tricky REM."
~~~

At this point, the local Cantata ***library db*** has (`comment` is not handled by the library):

~~~.text
$ sqlite3 -column -header $HOME/.local/share/cantata/library/localhost_6600.sql "select album,disc,track,origYear, file from songs where albumId LIKE '%ALBUM-C1-HCP%' AND title LIKE '%TRACK 01%'"

album       disc        track       origYear    file                                      
----------  ----------  ----------  ----------  ------------------------------------------
            2           1           1999        cue:///TEST-ALBUM-C1-HCP/audiocd.cue?pos=0
~~~

and the Library view shows correctly *Disc number* and *Original Year* for this track:

![0-test-library-view](https://user-images.githubusercontent.com/35839588/57156610-e806e500-6dcd-11e9-97b7-542a5c379bcc.png)

Now, when put the track in the ***play queue***, the debug log shows `MPDParseUtils::parseSongs` and `MPDParseUtils::parseSong` are executed again, and **around here seems the trouble had happened**:

~~~.text
2019-05-03 11:44:45 - MPDParseUtils *** Song MPDParseUtils::parseSongs
2019-05-03 11:44:45 - MPDParseUtils >>> data: "file: TEST-ALBUM-C1-HCP/audiocd.flac\nRange: 0.000-237.800\nLast-Modified: 2018-02-04T01:19:33Z\nGenre: Classical\nDate: 2007-05-08\nComment: Remark2 - A tricky REM.\nComposer: H Composer\nAlbumArtist: H Performer\nAlbum: ALBUM-C1-HCP\nTrack: 01\nArtist: Performer T1\nTitle: TRACK 01\nTime: 238\nduration: 237.800\nPos: 0\nId: 3\nOK\n"
...
2019-05-03 11:44:45 - MPDParseUtils Song MPDParseUtils::parseSong -- song.
2019-05-03 11:44:45 - MPDParseUtils    file: "TEST-ALBUM-C1-HCP/audiocd.flac"
2019-05-03 11:44:45 - MPDParseUtils    genre: "Classical" | album: "ALBUM-C1-HCP" | year: 2007 | originalYear: 0
2019-05-03 11:44:45 - MPDParseUtils    discNo: 0 | trackNo: 1 | title: "TRACK 01" | duration:  238
2019-05-03 11:44:45 - MPDParseUtils    artist: "Performer T1" | performer: "" | composer: "H Composer"
2019-05-03 11:44:45 - MPDParseUtils    comment: "Remark2 - A tricky REM."
...
~~~

Please, note how `data` has no `date` or `origYear`, and how `comment` has changed its value. Consequently, it seems to me the reason why in the Play Queue view there is no `date` or `origYear`:

![1-test-queue-view](https://user-images.githubusercontent.com/35839588/57156632-f3f2a700-6dcd-11e9-8f9f-152f64282366.png)

and why the same is in the Info/Now Playing view:

![2-test-nowplaying-view](https://user-images.githubusercontent.com/35839588/57156649-fbb24b80-6dcd-11e9-95ce-25c6bd850860.png)

Now, when playing this track I can see:

~~~.text
$ mpc -f "%title%,  %disc%, %track%, %originaldate%, %comment%, %file%" current

TRACK 01,  , 01, %originaldate%, Remark2 - A tricky REM., TEST-ALBUM-C1-HCP/audiocd.flac
~~~

where `disc` is missing, `%originaldate%` seems not be handled at all, and `%comment%` is the last one of the multi-line comment in the CUE file.

So I seem to understand that, **in the Play Queue view and in the Info/Now Playing view, Cantata gets metadata direcly from the MPD database**, and the problem is that MPD lacks in the support of CUE files.

This seems to be confirmed when, if modifying the (syntactically valid) CUE file rows:

~~~.text
REM ORIGINALYEAR "1999"
REM DISCNUMBER "2"
REM "Remark1 - A standard REM."
REM COMMENT "Remark2 - A tricky REM."
~~~

with those (syntactically valid) commands:
~~~.text
REM ORIGINALYEAR "1999"
REM ORIGINALDATE "1999"
REM ORIGINALDATE "1999-01-01"
REM DISCNUMBER "2"
REM DISC "2"
REM "Remark1 - A standard REM."
REM COMMENT "Remark2 - A tricky REM."
REM COMMENT "Remark1 + Remark2."
~~~

this results in MPD as:

~~~.text
$ mpc -f "%title%,  %disc%, %track%, %originaldate%, %comment%, %file%" current

TRACK 01,  2, 01, %originaldate%, Remark1 + Remark2., TEST-ALBUM-C1-HCP/audiocd.flac
~~~

and the same being diplayed by the *Info/Now Playing* view in Cantata!


What do you think about it? Is this analysis correct???

There can be a (possibly simple) solution with Cantata, which "corrects" the shortcomings with CUE files in MPD ?

Please tell me if I should open a new issue for this.

### in conclusion...

The zipped attachment contains the `cuefile.cpp` and `songview.cpp` codefiles, modified as described above.

[codefiles.zip](https://github.com/CDrummond/cantata/files/3143071/codefiles.zip)

Using these code files, I did a lot of tests, both with CUE files that I "created" specifically for this, and with the numerous CUE files in my collection: no problems encountered (except for the "strange" behavior described above, but it's a different matter).

I hope that the work I have done can in fact be a useful improvement for CUE support, accepted by you and included in Cantata. Let me know...

Please excuse me for not being perhaps as brief as I wanted.

Thank you.

Regards

**Comment by CDrummond** [comment]
Wow! That's a lot of text. I'll look at the code changes later, as I have no time right now. But anyway, thanks for doing this - I'll try not to revert too much ;)

A couple of points:

1. "According to you, Album Artist (which is not used) no longer exists anywhere." ???? Um? Really? If so, I'm wrong. There should be an "Album Artist". Not sure where you think I'm stating this should not be the case.

2. No please, I dont want the queue logic parsing cue files - that would be way too slow for lots of tracks. IF you want this info there, then it should come from MPD - and would benefit all clients,

**Comment by nousernouser** [comment]
>     1. "According to you, Album Artist (which is not used) no longer exists anywhere." ???? Um? Really? If so, I'm wrong. There should be an "Album Artist". Not sure where you think I'm stating this should not be the case.

Sorry, was a typo... I mean _Album Composer_; and "anywhere" was referred and limited to `cuefile.cpp`.  

Anyway, it does not seems to be necessary _Album Artist_ to be handled during parsing CUE files in `cuefile.cpp`. In fact:
- artist (mpd) is related to PERFORMER (cue)
- in `cuefile.cpp` we should use _Artist_ instead of _Album Artist_ because Cantata uses somewhere else ArtistOrComposer() to manage the album grouping ... and that works as expected...

>     2. No please, I dont want the queue logic parsing cue files - that would be way too slow for lots of tracks. If you want this info there, then it should come from MPD - and would benefit all clients,

I apologize but I'm not sure I understand what you mean ...

The logic is:

- Each cue file is read line by line; 
- for each line the tag value (when present) is stored in a simple string (eg: album = cmdVal); 
- when the end of the file is reached, the current track data is stored adding a new element at the end of a QVector -- which helds only the found ("tag", "tagvalue") QHash pairs which are created on the fly;
- the vector initially is empty and is resized adding a new element when storing a track data (so at the end of the parsing it has many elements as the number of tracks).

In such a parsing, there is no the logic of a queue.

At the most, it can be said that the vector which is used behaves in practice as a list. But a list is what you had in the "original" version of `cuefile.cpp`, where was `QList <CueEntry> entries` involving, for each CUE file, to store within a Qlist a CueEntry structure for each track .

From what I've read in the documentation, Qt now recommends using QVector rather than QList (see: <https://doc.qt.io/qt-5/qlist.html#details>), and also recommends using QHash to manage pairs of values ​​(see: <https://doc.qt.io/qt-5/qhash.html#details>): this should come to better performance.

The impression I had with the tests done with my wide (really wide) audio collection (CUE files) accords with this: the performances when performing at "Refresh Database" seems to have been improved.

Maybe some other user can do tests? 

>     3. Cantata does not add "comment" to its local DB - as there is no point. The local DB is use for library loading, and the library code does not use comment.

So it seems my analisys regarding the "strange" behaviour was correct...


Thank you. Then I wait ... 

Let me know if you need clarification on what I have proposed.

Regards

**Comment by nousernouser** [comment]
*UPDATE* (seems github has unflagged my account...)

Here are the latest cuefile.cpp and songview.cpp (which do not were attached in the previous comment I have deleted. Please do *not* consider the files I sent before. 

[codefiles.zip](https://github.com/CDrummond/cantata/files/3154819/codefiles.zip)

Please note:
- Album Artist is managed;
- parsing flow has been reviewed optimizing memory allocation and execution speed (following recommends in the Qt documentation -- please have a look at the detailed comments I have put inside the code). Parsing performance seems to be good!
- no more "strange" behaviour in the Info/Now Playing view...

Not completely sure the changes also solve the #1479 issue... I successfully did some test with my CUEs, but maybe other tests should be done being possible a lot of different FILE-TRACKs combinations... 

Hope you will accept the work done, including it in Cantata. Please let me know about it.
Regards

**Comment by CDrummond** [comment]
I've applied 99.9% of the cue file changes - made a small mod to isVariousArtists()
Did not like the changes to songview.cpp, so not applied.

Can this now be closed?

**Comment by nousernouser** [comment]
> I've applied 99.9% of the cue file changes - made a small mod to isVariousArtists()

Great. Thank you.

> Did not like the changes to songview.cpp, so not applied.

Ok. Just two things about it.

1. I have seen that nevertheless you have made some changes, and now there seems to be a problem in the display of the metadata (see image attached) which should be fixed: they are listed in reverse order.
![songview-metadata](https://user-images.githubusercontent.com/35839588/57585185-dcf63800-74e4-11e9-8a37-e19a168715f8.png)

2. In CUE files the 'COMMENT' tag is sometimes used to insert additional metadata that is not covered by standard commands (ie the release description, audio sources information etc.), using newlines as separators (which are allowed by the standard) . It's a very useful trick (I also have several CUE files that use it) ... For this reason I think it would be appropriate to handle newlines in the 'comment' tag in order to have this trick working (in a similar way as I tried to do it in the proposed code). I would greatly appreciate it. Thanks.

> Can this now be closed?

For sure regarding cuefile.cpp!  
PS: I did other tests with multiple tracks/files (issue #1479) in different combinations, and all seems to be working well.

Maybe not for s.cpp ... 

Thanks a lot.
Regards

**Comment by nousernouser** [comment]
> 1. I have seen that nevertheless you have made some changes, and now there seems to be a problem in the display of the metadata (see image attached) which should be fixed: they are listed in reverse order.

The attached patch should fix the "reverse metadata display order" issue (supposing it was a typo, also changed 'addIfNotEmty' to 'addIfNotEmpty').

Regards

[songview.cpp.patch.zip](https://github.com/CDrummond/cantata/files/3172481/songview.cpp.patch.zip)

**Comment by nousernouser** [comment]
Thank you.
Regards
